vcd/resource_vcd_independent_disk.go

```diff
@@ -173,33 +207,204 @@ func resourceVcdIndependentDiskCreate(d *schema.ResourceData, meta interface{})
+		return diag.Errorf("unable to find disk with href %s: %s", diskHref, err)
 	}
 
 	d.SetId(disk.Disk.Id)
 
-	return resourceVcdIndependentDiskRead(d, meta)
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
 }
 
-func resourceVcdIndependentDiskRead(d *schema.ResourceData, meta interface{}) error {
+func resourceVcdIndependentDiskUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	vcdClient := meta.(*VCDClient)
+
+	if d.HasChanges("size_in_mb", "storage_profile", "description") {
+		_, vdc, err := vcdClient.GetOrgAndVdcFromResource(d)
+		if err != nil {
+			return diag.Errorf(errorRetrievingOrgAndVdc, err)
+		}
+
+		storageProfileValue := d.Get("storage_profile").(string)
+		var storageProfileRef *types.Reference
+
+		if storageProfileValue != "" {
+			storageReference, err := vdc.FindStorageProfileReference(storageProfileValue)
+			if err != nil {
+				return diag.Errorf("error finding storage profile %s", storageProfileValue)
+			}
+			storageProfileRef = &types.Reference{HREF: storageReference.HREF}
+		}
+
+		disk, err := vdc.GetDiskById(d.Id(), true)
+		if err != nil {
+			return diag.Errorf("error fetching independent disk: %s", err)
+		}
+
+		diskAttachedVmsHrefs, err := disk.GetAttachedVmsHrefs()
+		if err != nil {
+			return diag.Errorf("error resourceVcdIndependentDiskUpdate faced issue fetching attached VMs")
+		}
+
+		// Add global lock for shared disks to avoid deadlock possibility when different independent shared disks used by same VMs
+		if len(diskAttachedVmsHrefs) > 1 {
+			lockSharedDiskOpsGlobally()
+			defer unlockSharedDiskOpsGlobally()
+		}
+		//lock VMs as another independent disk resource can be doing update with same VM
+		lockVms(diskAttachedVmsHrefs)
+		defer unlockVms(diskAttachedVmsHrefs)
+
+		diskDetailsForReAttach, diagErr := detachVms(vcdClient, disk, diskAttachedVmsHrefs)
+		if diagErr != nil {
+			return diagErr
+		}
+
+		err = disk.Refresh()
+		if err != nil {
+			return diag.Errorf("error resourceVcdIndependentDiskUpdate error refreshing independent disk: %s", err)
+		}
+
+		disk.Disk.SizeMb = int64(d.Get("size_in_mb").(int))
+		disk.Disk.Description = d.Get("description").(string)
+		if storageProfileRef != nil {
+			disk.Disk.StorageProfile = storageProfileRef
+		}
+
+		task, err := disk.Update(disk.Disk)
+		if err != nil {
+			return diag.Errorf("error updating independent disk: %s", err)
+		}
+
+		err = task.WaitTaskCompletion()
+		if err != nil {
+			return diag.Errorf("error waiting to finish updating of independent disk: %s", err)
+		}
+
+		diagErr = attachBackVms(vcdClient, disk, diskDetailsForReAttach, diskAttachedVmsHrefs)
+		if diagErr != nil {
+			return diagErr
+		}
+
+	}
+	return resourceVcdIndependentDiskRead(ctx, d, meta)
+}
+
+func lockSharedDiskOpsGlobally() {
+	vcdMutexKV.kvLock(globalIndependentDiskLockKey)
+}
+
+func unlockSharedDiskOpsGlobally() {
+	vcdMutexKV.kvUnlock(globalIndependentDiskLockKey)
+}
+
+func lockVms(sliceOfVmsHrefs []string) {
```

Comment:
Please add a comment about the purpose of this func.

Reply:
added